packages/flutter/test/material/material_state_test.dart

```diff
@@ -0,0 +1,378 @@
+                  return const Text(hoverText);
+                if (states.contains(MaterialState.focused))
+                  return const Text(focusText);
+                if (states.contains(MaterialState.pressed))
+                  return const Text(pressText);
+                return const Text(defaultStateText);
+              },
+            ),
+          ),
+        ),
+      );
+    }
+
+    // Default state.
+    await tester.pumpWidget(entireWidget());
+    expect(find.text(defaultStateText), findsOneWidget);
+    expect(find.text(disabledText), findsNothing);
+    expect(find.text(selectedText), findsNothing);
+    expect(find.text(errorText), findsNothing);
+    expect(find.text(hoverText), findsNothing);
+    expect(find.text(focusText), findsNothing);
+    expect(find.text(pressText), findsNothing);
+
+    // Disabled state.
+    await tester.pumpWidget(entireWidget(disabled: true));
+    expect(find.text(defaultStateText), findsNothing);
+    expect(find.text(disabledText), findsOneWidget);
+    expect(find.text(selectedText), findsNothing);
+    expect(find.text(errorText), findsNothing);
+    expect(find.text(hoverText), findsNothing);
+    expect(find.text(focusText), findsNothing);
+    expect(find.text(pressText), findsNothing);
+
+    // Re-enabled.
+    await tester.pumpWidget(entireWidget(disabled: false));
+    expect(find.text(defaultStateText), findsOneWidget);
+    expect(find.text(disabledText), findsNothing);
+
+    // Selected state.
+    await tester.pumpWidget(entireWidget(selected: true));
+    expect(find.text(defaultStateText), findsNothing);
+    expect(find.text(disabledText), findsNothing);
+    expect(find.text(selectedText), findsOneWidget);
+
+    // Un-select.
+    await tester.pumpWidget(entireWidget(selected: false));
+    expect(find.text(defaultStateText), findsOneWidget);
+    expect(find.text(disabledText), findsNothing);
+    expect(find.text(selectedText), findsNothing);
+
+    // Error state.
+    await tester.pumpWidget(entireWidget(error: true));
+    expect(find.text(defaultStateText), findsNothing);
+    expect(find.text(errorText), findsOneWidget);
+
+    // Undo error.
+    await tester.pumpWidget(entireWidget(error: false));
+    expect(find.text(defaultStateText), findsOneWidget);
+    expect(find.text(errorText), findsNothing);
+
+    // Hover state.
+    await tester.pumpWidget(entireWidget());
+    final TestGesture mouse = await tester.createGesture(kind: PointerDeviceKind.mouse);
+    await mouse.moveTo(tester.getCenter(find.byType(Text)));
+    await tester.pump(); // Register hover.
+    await tester.pump(); // Update Based on hover.
+    expect(find.text(defaultStateText), findsNothing);
+    expect(find.text(hoverText), findsOneWidget);
+
+    // Un-hover.
+    await mouse.moveTo(tester.getTopRight(find.byType(MaterialApp)));
+    await tester.pump(); // Register hover.
+    await tester.pump(); // Update Based on hover.
```

Comment:
So you have to do the double pump then? Did you find out why this is necessary?

Reply:
I believe hover and focus always require double pumps, the issue I was having before was the fact that buttons were now requiring an additional third pump for press, which I fixed by having the gesture detector be optional.

Reply:
Yes as I understand it hover, focus, and press always require one frame to register and then one frame to update. My problem was that I was accidentally _adding_ an additional frame to buttons, I fixed that by making the `GestureDetector` optional in `MaterialStateBuilder`.